flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ExecutionGraphCheckpointCoordinatorTest.java

```diff
@@ -131,9 +142,82 @@ public void testShutdownCheckpointCoordinatorOnFinished() throws Exception {
 		assertThat(storeShutdownFuture.get(), is(JobStatus.FINISHED));
 	}
 
+	/**
+	 * The case is designed to check the race condition between {@link ExecutionGraph} and
+	 * {@link CheckpointCoordinator}. There should be no checkpoint accepted after
+	 * {@link CheckpointFailureManager} decides to fail the {@link ExecutionGraph}.
+	 */
+	@Test
+	public void testNoCheckpointAcceptedWhileFailingExecutionGraph() throws Exception {
+		CheckpointIDCounter counter = new TestingCheckpointIDCounter(new CompletableFuture<>());
+
+		TestingCompletedCheckpointStore store = new TestingCompletedCheckpointStore(new CompletableFuture<>());
+
+		final ManuallyTriggeredScheduledExecutor mainThreadExecutor = new ManuallyTriggeredScheduledExecutor();
+		ExecutionGraph graph = createExecutionGraphAndEnableCheckpointing(
+			counter,
+			store,
+			new ComponentMainThreadExecutorServiceAdapter(mainThreadExecutor, Thread.currentThread()));
+
+		final CheckpointCoordinator checkpointCoordinator = graph.getCheckpointCoordinator();
+
+		assertThat(checkpointCoordinator, Matchers.notNullValue());
+		assertThat(checkpointCoordinator.isShutdown(), is(false));
+
+		graph.scheduleForExecution();
+		ExecutionGraphTestUtils.switchToRunning(graph);
+
+		// trigger a normal checkpoint which would be declined then
+		checkpointCoordinator.triggerCheckpoint(System.currentTimeMillis(), true);
+		mainThreadExecutor.triggerAll();
+		final long checkpointId = checkpointCoordinator.getPendingCheckpoints().keySet().iterator().next();
```

Comment:
~Why is this checkpoint going to be declined? Maybe expand the comment with a bit of explanation?~

~If it's failing, maybe add an assertion that it has failed? Or just skip this code block if it's not necessary for the test as a whole?~

edit: 
replace
```
// trigger a normal checkpoint which would be declined then
```
with
```
// trigger a normal checkpoint which we will fail/decline later
```
to avoid confusion as mine above? :)

Reply:
sure